### ISSUE TITLE
WASM middleware state not kept

### ISSUE THREAD

**Issue by darkweak** [traefik/traefik#10953]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you do?

I'm actually looking at the WASM support for plugins in Træfik as @emilevauge mentioned but with this minimal snippet, we can see that the init is called when I send an HTTP request to Træfik

```go
package main

import (
	"github.com/http-wasm/http-wasm-guest-tinygo/handler"
	"github.com/http-wasm/http-wasm-guest-tinygo/handler/api"
)

func main() {
	handler.Host.Log(api.LogLevelWarn, "INIT!!!!!")

	handler.HandleRequestFn = func(r1 api.Request, r2 api.Response) (next bool, reqCtx uint32) {
		return true, 0
	}
}
```

I would like to know if that's the normal behavior or if that's a bug.  
IMHO loading WASM plugins as stateless and reconfigure them each time you go through the middleware is a performance issue.

### What did you see instead?

```
traefik-1  | 2024-07-30T21:25:08Z WRN github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > INIT!!!!! entryPointName=http middlewareName=souin@file middlewareType=wasm routerName=whoami@file
traefik-1  | 2024-07-30T21:25:08Z DBG github.com/traefik/traefik/v3/pkg/server/service/loadbalancer/wrr/wrr.go:196 > Service selected by WRR: b903963fe23550fa
traefik-1  | 192.168.65.1 - - [30/Jul/2024:21:25:08 +0000] "GET / HTTP/1.1" 200 373 "-" "-" 1 "whoami@file" "http://whoami" 9ms
traefik-1  | 2024-07-30T21:25:09Z WRN github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > INIT!!!!! entryPointName=http middlewareName=souin@file middlewareType=wasm routerName=whoami@file
traefik-1  | 2024-07-30T21:25:09Z DBG github.com/traefik/traefik/v3/pkg/server/service/loadbalancer/wrr/wrr.go:196 > Service selected by WRR: b903963fe23550fa
traefik-1  | 192.168.65.1 - - [30/Jul/2024:21:25:09 +0000] "GET / HTTP/1.1" 200 373 "-" "-" 2 "whoami@file" "http://whoami" 4ms
traefik-1  | 2024-07-30T21:25:09Z DBG github.com/traefik/traefik/v3/pkg/server/service/loadbalancer/wrr/wrr.go:196 > Service selected by WRR: b903963fe23550fa
traefik-1  | 192.168.65.1 - - [30/Jul/2024:21:25:09 +0000] "GET / HTTP/1.1" 200 373 "-" "-" 3 "whoami@file" "http://whoami" 2ms
```

### What version of Traefik are you using?

Traefik version 3.1.0 built on 2024-07-15T14:44:04Z

### What is your environment & configuration?

```yaml
# traefik.yml
# ...
experimental:
  localPlugins:
    souin:
      moduleName: github.com/darkweak/souin/plugins/traefik/wasm
```

```yaml
# .traefik.yml
displayName: Souin
type: middleware
runtime: wasm
import: github.com/darkweak/souin/plugins/traefik/wasm
summary: 'Souin is a powerfull cache system as fast as Varnish but easier to configure'
```

### If applicable, please paste the log output in DEBUG level

```
traefik-1  | 2024-07-30T21:25:08Z WRN github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > INIT!!!!! entryPointName=http middlewareName=souin@file middlewareType=wasm routerName=whoami@file
traefik-1  | 2024-07-30T21:25:08Z DBG github.com/traefik/traefik/v3/pkg/server/service/loadbalancer/wrr/wrr.go:196 > Service selected by WRR: b903963fe23550fa
traefik-1  | 192.168.65.1 - - [30/Jul/2024:21:25:08 +0000] "GET / HTTP/1.1" 200 373 "-" "-" 1 "whoami@file" "http://whoami" 9ms
traefik-1  | 2024-07-30T21:25:09Z WRN github.com/traefik/traefik/v3/pkg/logs/wasm.go:31 > INIT!!!!! entryPointName=http middlewareName=souin@file middlewareType=wasm routerName=whoami@file
traefik-1  | 2024-07-30T21:25:09Z DBG github.com/traefik/traefik/v3/pkg/server/service/loadbalancer/wrr/wrr.go:196 > Service selected by WRR: b903963fe23550fa
traefik-1  | 192.168.65.1 - - [30/Jul/2024:21:25:09 +0000] "GET / HTTP/1.1" 200 373 "-" "-" 2 "whoami@file" "http://whoami" 4ms
traefik-1  | 2024-07-30T21:25:09Z DBG github.com/traefik/traefik/v3/pkg/server/service/loadbalancer/wrr/wrr.go:196 > Service selected by WRR: b903963fe23550fa
traefik-1  | 192.168.65.1 - - [30/Jul/2024:21:25:09 +0000] "GET / HTTP/1.1" 200 373 "-" "-" 3 "whoami@file" "http://whoami" 2ms
```

**Comment by juliens** [comment]
Hello @darkweak ,

For the wasm plugin middlewares, it works with a pool so it's possible that when the second request happens, the first middleware was not yet back in the pool.

You can see in your logs that the third request does not call the "INIT".

**Comment by darkweak** [comment]
Hey @juliens, do you have clues to prevent that?

**Comment by juliens** [comment]
Do you have a reason to prevent this? 

It's designed to handle multiple requests in parallel. The pool should be populated over time, and once it's filled, it shouldn't call INIT anymore.

**Comment by darkweak** [comment]
I'm migrating the HTTP cache from go middleware to the WASM middleware (due to limitations from yægi).  
It tries to recreate more and more instances, so I think the pool resize itself, kill the previous instances and recreate the middleware when a new request comes.

**Comment by juliens** [comment]
The pool may remove some instances from the pool, I believe to refresh old instances, but it doesn't instantiate for each requests.

With your example middleware:
With an hey command
 703091 responses
I only have 
322 INIT

**Comment by darkweak** [comment]
That means, it could create so much connections to the storage provider (e.g. redis/etcd) and there is no way to avoid reinitialization.  
If there is a reset when using an in-memory storage, that will result a miss and reach the upstream server too often.

**Comment by juliens** [comment]
Yes, the current implementation of the WASM plugin middleware in Traefik does have some limitations, particularly regarding the instance pool and initialization behavior you're observing. Each middleware instance may get reinitialized more frequently than desired, especially with external connections like Redis or etcd and with shared-memory cache.

However, as there is a pool, the situation is not as dramatic as it might sound. The pool is designed to manage multiple requests in parallel and, over time, it should stabilize, reducing the frequency of initialization calls. This helps mitigate the performance impact to some extent.

Moreover, there are ways to improve this situation that we could try in the futur:

Shared In-Memory Objects: One potential improvement is to add support for shared objects in memory. This would allow instances to share certain data, reducing the need to reinitialize connections or cache data on each request.

Enhanced Pool Management: Enhancing the pool's management strategies to improve instance lifetime can help mitigate excessive reinitialization. This includes better handling of pool resizing and more stable instance retention policies.

Clean callback: In the Yaegi middleware, when a middleware is not used anymore, the context used in the instanciation is Done, it's not the case in wasm plugin, so it could be a good idea to have a callback to free resource like ecxternal connection to avoid infinite connection.

**Comment by ttys3** [comment]
> Shared In-Memory Objects:

`Shared In-Memory Objects:` is important.

with wasm plugin enabled, when concurrent requests comes, the memory increases very crazy.

![Image](https://github.com/user-attachments/assets/35c04356-303d-458a-97eb-b9d1d30e7046)
